### ISSUE TITLE
Return null or exception on EOF (end-of-file)

### ISSUE THREAD

**Issue by kanaka** [anseki/readline-sync#36]
There does not appear to currently be a way to distinguish between end-of-file (e.g. Ctrl-D on Linux) and an empty line. I suggest returning a null when EOF is encountered.

**Comment by anseki** [comment]
Hi @kanaka, thank you for your suggestion.
readlineSync reads single line. Therefore it finishes reading when EOF or line-break was input.

**Comment by kanaka** [comment]
@anseki yes, but if you are trying to use readlineSync in a loop to build an interactive REPL, then you need to be able to distinguish between an empty line (user pressed enter) and an EOF (user pressed Ctrl-D and wants to quit).

**Comment by anseki** [comment]
I see.
If you want to make your script implement CLI, [promptCLLoop](https://github.com/anseki/readline-sync#promptclloop) may help you.
Unfortunately, it doesn't support `EOF`, but it supports a command to exit from CLI.
For example:

```js
readlineSync.promptCLLoop({
  command1: arg1 =>         { console.log(`command1 does something with <${arg1}>...`); },
  command2: (arg1, arg2) => { console.log(`command2 does something with <${arg1}> and <${arg2}>...`); },
  bye: () => true
});
console.log('Exited');
```

```
> command1 foo
command1 does something with <foo>...
> command2 bar "B A Z"
command2 does something with <bar> and <B A Z>...
> bye
Exited
```

CLIs that are commonly used such as REPL of Node.js support a command to exit from it (e.g. `.exit` command of REPL of Node.js). `SIGINT` signal (`^C`) also is used in many CLIs.
You have to use a command or `SIGINT` to exit from CLI in current version of `readlineSync.promptCLLoop`.
I think that supporting `EOF` is not easy because it must distinguish the "request to exit" and "real `EOF`" (i.e. end of data user input). And the changing it affects many things (e.g. API, internal functions, etc.).
Anyway, I may consider it in future version.
Thank you.

**Comment by kanaka** [comment]
Ctrl-D is pretty much the universal way to terminate any sort of input prompt under Linux/UNIX. Even simple C programs that just use `read` or `scanf` treat Ctrl-D as EOF. Encountering a program that takes interactive input but does not accept Ctrl-D is a jarring experience for Linux/UNIX command line users. One additional advantage of Ctrl-D over Ctrl-C is that with Ctrl-D, a application can trivially detect that EOF has been encountered and do cleanup tasks (delete temp files, etc). In order to cleanup after a Ctrl-C the application usually has to trap signals properly (non-trivial) in order to do cleanup.

Well behaved full fledged REPL programs on Linux/UNIX do not tend to exit with Ctrl-C. Rather, Ctrl-C cancels whatever is currently typed (if anything) and presents a fresh prompt. Ctrl-D on the other hand will do nothing if there is typed but not send text (user has typed something but not yet pressed enter). This is true for python, irb (ruby), lein repl (clojure), rlwrap (readline wrapper), etc. Even the built-in node "repl" module has the normal Ctrl-C and Ctrl-D behavior. The built-in node "readline" module doesn't trap Ctrl-C, but even it still has standard Ctrl-D behavior like the other REPLs (Ctrl-D exits when no input, does nothing with pending input). It's true that most REPLs have a specific command that will exit, but it's usually specific to that particular REPL. Ctrl-D is pretty much universal and you don't have to figure it out per REPL (I always use Ctrl-D and I'm sure that's true of most Linux/UNIX REPL users).

I agree that this is non-trivial to implement. I took a look at the code a few months ago to see if I could quickly add it, but eventually just settled on calling readline/libedit directly via FFI. But it would be nice to have a more cross-platform way of doing synchronous readline without resorting to a compiled node module (FFI).

**Comment by anseki** [comment]
As I said, readlineSync accepts `EOF`. But it only finishes reading when EOF was input.
Yes, I also think that if the `EOF` can be used to exit the loop, it's useful.
I may consider it in future version.
Thank you for your suggestion.

**Comment by kanaka** [comment]
BTW, this doesn't cover the edge cases (e.g. the external program calls), but this patch works in my normal case at least:

```patch
diff --git a/lib/readline-sync.js b/lib/readline-sync.js
index 455ee22..79dd69a 100644
--- a/lib/readline-sync.js
+++ b/lib/readline-sync.js
@@ -386,6 +386,13 @@ function _readlineSync(options) {
           return;
         }
       }
+
+      // EOF
+      if (readSize === 0 && input.length === 0) {
+          input = null;
+          return;
+      }
+
       chunk = readSize > 0 ? buffer.toString(options.encoding, 0, readSize) : '\n';
 
       if (chunk && typeof (line = (chunk.match(/^(.*?)[\r\n]/) || [])[1]) === 'string') {
```

**Comment by anseki** [comment]
Thank you for the test.
As you said, your code doesn't cover other situations such as raw-mode, `EOF` in Windows (`Ctrl+Z`), etc.
And also, as I said, this changing makes many functions fail because those functions expect a `string` to be passed.

**Comment by monis256** [comment]
While we the topic is on EOF, is there a way to capture only the enter key or limiting the user to press only enter @anseki

**Comment by anseki** [comment]
Hi @monis256,

Sorry, my English is very poor, and I could not understand your comment.
When the user presses only Enter key, `EOF` is not input, and readlineSync only finishes the reading.
Therefore, readlineSync returns `''` when the user presses only Enter key.
How do you use `EOF` for Enter key?

**Comment by monis256** [comment]
Sorry for the confusion. Let me rephrase. While using question/prompt/keyIN methods can i limit the user input to only "Enter key". I want to prompt "Press Enter to continue" and only "Enter key" should be allowed. Is this possible?

**Comment by anseki** [comment]
I see.
If you want to make the running of your script pause and let the user say "I'm ready", [keyInPause method](https://github.com/anseki/readline-sync#utility_methods-keyinpause) may help you.
This can accept any key or only Space bar (or another key) instead of Enter key.

**Comment by monis256** [comment]
Thank you. I tried that and I'm probably going with it. But i was wondering if "Enter Key" can be used somehow, because chances are that the user might hit the "Enter key" before anything else to continue.

**Comment by anseki** [comment]
As I said, readlineSync finishes the reading when the user presses Enter key.
Therefore, If you want to use Enter key for that, you can ignore the return value.

For example:

```js
readlineSync.question('Hit Enter key to continue.');
// Do something...
```

If you want to hide string that was input by the user, you can use `hideEchoBack` option.

For example:

```js
readlineSync.question('Hit Enter key to continue.', {hideEchoBack: true, mask: ''});
// Do something...
```

**Comment by monis256** [comment]
Oh the ``` hideEchoBack ``` option helped. Thank you.

**Comment by monis256** [comment]
Hi i know this is not perfect place to ask questions but i did have a some questions to you regarding readline-sync. what is the best way to contact you.

**Comment by anseki** [comment]
@monis256,
You can make a new Issue thread for the question by clicking "New Issue" button at top of this page.

**Comment by anseki** [comment]
readlineSync provide the raw input data via `getRawInput` method to do something with control characters, instead of controlling behavior.
I think this is better, because the cases of using control characters are not many.
